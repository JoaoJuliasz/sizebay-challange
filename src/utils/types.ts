export type FilterType = 'done' | 'pending' | ''

export type Items = {
    value: string,
    status: 'done' | 'pending'
    itemIndex: number
}