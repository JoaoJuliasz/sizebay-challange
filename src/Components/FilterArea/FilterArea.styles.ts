//Importing libs
import styled from 'styled-components'
import { Button } from 'antd'
import { SearchOutlined } from '@ant-design/icons';

type ButtonProps = {
    selectedButton: boolean
}

const activeButtonStyle = `
        border: 1px solid #4DA6B3;
        background: #F7F7F8 0% 0% no-repeat padding-box;
        color: #4DA6B3;
    `

export const FilterAreaContainer = styled.div`
    display: flex;
    margin: 10px auto;
    width: 100%;
    justify-content: space-around;

    @media only screen and (max-width: 620px) {
        flex-direction: column-reverse;
        align-items: center;
    }
`

export const FilterButtonsContainer = styled.div`
    display: flex;

    @media only screen and (max-width: 620px) {
        justify-content: end;
        margin: 5px;
        width: 90%;
        max-width: 510px;

    }
`

export const SearchAreaContainer = styled.div`
    position: relative;

    @media only screen and (max-width: 620px) {
        width: 90%;
        max-width: 500px;
    }
`

export const SearchArea = styled.input`
    min-width: 350px;
    padding: 10px;
    outline: none;
    border: 1px solid #dbdbdb;
    border-radius: 5px;

    @media only screen and (max-width: 620px) {
        width: 100%;
        min-width: auto;
    }
`

export const SearchIcon = styled(SearchOutlined)`
    position: absolute;
    top: 30%;
    right: 10px;
    font-size: 18px;
`

export const FilterButton = styled(Button) <ButtonProps>`
    margin: 0 5px;
    padding: 5px 10px;
    border-radius: 23px;
    color: #848484;
    height: 35px;

    ${props => props.selectedButton && activeButtonStyle}

    &:hover, &:active, &:focus {
       ${activeButtonStyle}
    }
`

