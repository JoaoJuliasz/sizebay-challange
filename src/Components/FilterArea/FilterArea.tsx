//Importing libs
import { Button } from 'antd'
import { CheckOutlined } from '@ant-design/icons';

//Importing styled-components
import { FilterAreaContainer, FilterButtonsContainer, SearchAreaContainer, SearchArea, FilterButton, SearchIcon } from './FilterArea.styles'

import { FilterType } from '../../utils/types'

type FilterAreaProps = {
    searchField: string
    setSearchField: (value: string) => void
    filterType: string
    setFilterType: (value: FilterType) => void
}

const FilterArea = ({ searchField, setSearchField, filterType, setFilterType }: FilterAreaProps) => {

    const handleClick = (value: FilterType) => {
        let newValue = filterType as FilterType
        if (value === filterType) {
            newValue = ''
            setFilterType(newValue)
        } else {
            newValue = value
            setFilterType(newValue)
        }
    }

    return (
        <FilterAreaContainer>
            <FilterButtonsContainer>
                <FilterButton id="done" onClick={() => handleClick('done')} selectedButton={'done' === filterType}
                    size="large" icon={'done' === filterType && <CheckOutlined />}>Done</FilterButton>
                <FilterButton id="pending" selectedButton={'pending' === filterType} icon={'pending' === filterType && <CheckOutlined />} onClick={() => handleClick('pending')} size="large">Pending</FilterButton>
            </FilterButtonsContainer>
            <SearchAreaContainer>
                <SearchArea id="search-filter" type="text" value={searchField} placeholder="Search items"
                    onChange={e => setSearchField(e.target.value)} />
                <SearchIcon />
            </SearchAreaContainer>
        </FilterAreaContainer>
    );
};

export default FilterArea;