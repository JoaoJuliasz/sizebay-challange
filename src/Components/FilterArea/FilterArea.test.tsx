import { shallow } from 'enzyme'
import FilterArea from './FilterArea'
import { Items } from '../../utils/types';

describe('FilterArea', () => {

    let mockSearchField = ''
    let mockFilterType = ''
    const mockSetSearchField = jest.fn();
    const mockSetFilterType = jest.fn();

    const mockItems:Items[] = [
        {value: 'item 1', status: 'done', itemIndex: 0},
        {value: 'item 2', status: 'pending', itemIndex: 1},
    ]

    const mockFilterItems = (type: 'value' | 'status', filter: string) => mockItems.filter((item: Items) => item[type]?.includes(filter))

    const filterAreaWrapper = shallow(<FilterArea searchField={mockSearchField} filterType={mockFilterType}
        setSearchField={mockSetSearchField} setFilterType={mockSetFilterType} />)

    it('expect to render FilterArea component', () => {
        expect(filterAreaWrapper).toMatchSnapshot()
    })

    it('expect to filter done items', () => {
        filterAreaWrapper.find('[id="done"]').simulate('click')
        expect(mockSetFilterType).toHaveBeenCalled()
        expect(mockFilterItems('status', 'done').length).toEqual(1)
    })
    it('expect to filter pending items', () => {
        filterAreaWrapper.find('[id="pending"]').simulate('click')
        expect(mockSetFilterType).toHaveBeenCalled()
        expect(mockFilterItems('status', 'pending').length).toEqual(1)
    })
    it('expect to filter items using search field', () => {
        filterAreaWrapper.find('[id="search-filter"]').simulate('change', {target: {value: 'item 1'}})
        expect(mockSetSearchField).toHaveBeenCalled()
        expect(mockFilterItems('value', 'item 1').length).toEqual(1)
    })
})