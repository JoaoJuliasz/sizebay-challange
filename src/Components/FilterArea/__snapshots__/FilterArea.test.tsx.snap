// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterArea expect to render FilterArea component 1`] = `ShallowWrapper {}`;
